src/libCrypto/EthCrypto.cpp

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (C) 2022 Zilliqa
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include "EthCrypto.h"
+
+#include <openssl/ec.h>  // for EC_GROUP_new_by_curve_name, EC_GROUP_free, EC_KEY_new, EC_KEY_set_group, EC_KEY_generate_key, EC_KEY_free
+#include <openssl/obj_mac.h>  // for NID_secp192k1
+#include <openssl/sha.h>      //for SHA512_DIGEST_LENGTH
+#include <ethash/keccak.hpp>
+#include <functional>
+#include <iostream>
+#include <memory>
+#include <string>
+
+using namespace std;
+
+// Inspiration from:
+// https://stackoverflow.com/questions/10906524
+// https://stackoverflow.com/questions/57385412/
+// https://medium.com/mycrypto/the-magic-of-digital-signatures-on-ethereum-98fe184dc9c7
+
+// Prefix signed txs in Ethereum with Keccak256("\x19Ethereum Signed
+// Message:\n32" + Keccak256(message))
+constexpr uint8_t prelude[] = {25,  69,  116, 104, 101, 114, 101, 117, 109,
+                               32,  83,  105, 103, 110, 101, 100, 32,  77,
+                               101, 115, 115, 97,  103, 101, 58,  10,  48};
+
+// https://stackoverflow.com/questions/57385412/
+void SetOpensslSignature(const std::string& sSignatureInHex, ECDSA_SIG* pSign) {
+  // Openssl uses raw pointers and macros for freeing but can force it into
+  // smart pointers methodology
+  std::unique_ptr<BIGNUM, std::function<void(BIGNUM*)>> rr(
```

Comment:
As much as I like unique_ptr, if the usage of this pointer is only within what otherwise would be a 5 line function, I see no harm from just using `BN_free()` at the end.  Because it will not throw. Much less cognitive load in pure C style.  I don't insist though. (To me it just looks more modern, but solves a non-existing problem).  Wouldn't something like this be easier on the reader?

```
std::string sSignatureR =
      sSignatureInHex.substr(0, sSignatureInHex.size() / 2);
  std::string sSignatureS = sSignatureInHex.substr(sSignatureInHex.size() / 2);
BIGNUM *r = nullptr, *s = nullptr;   // Well known r, s numbers
BN_hex2n(&r, sSignatureR.c_str());
BN_hex2n(&s, sSignatureS.c_str());
ECDSA_SIG_set0(pSign, r, s);
BN_free(r);
BN_free(s);
```

Reply:
There is always a way your code could throw, even just when allocating memory :) I would opt to leave it in although I agree that it looks horrible. Perhaps we can get a second opinion from @CSideSteve @KaustubhShamshery or @moboware

Reply:
ok as I said I don't insist here.